command/agent/consul/check_watcher.go

```diff
@@ -0,0 +1,278 @@
+				c.allocID, c.taskName, c.checkName)
+			c.unhealthyStart = time.Time{}
+		}
+		return
+	}
+	switch status {
+	case api.HealthCritical:
+	case api.HealthWarning:
+		if c.ignoreWarnings {
+			// Warnings are ignored, reset state and exit
+			healthy()
+			return
+		}
+	default:
+		// All other statuses are ok, reset state and exit
+		healthy()
+		return
+	}
+
+	if now.Before(c.graceUntil) {
+		// In grace period exit
+		return
+	}
+
+	if c.unhealthyStart.IsZero() {
+		// First failure, set restart deadline
+		if c.timeLimit != 0 {
+			c.logger.Printf("[DEBUG] consul.health: alloc %q task %q check %q became unhealthy. Restarting in %s if not healthy",
+				c.allocID, c.taskName, c.checkName, c.timeLimit)
+		}
+		c.unhealthyStart = now
+	}
+
+	// restart timeLimit after start of this check becoming unhealthy
+	restartAt := c.unhealthyStart.Add(c.timeLimit)
+
+	// Must test >= because if limit=1, restartAt == first failure
+	if now.Equal(restartAt) || now.After(restartAt) {
+		// hasn't become healthy by deadline, restart!
+		c.logger.Printf("[DEBUG] consul.health: restarting alloc %q task %q due to unhealthy check %q", c.allocID, c.taskName, c.checkName)
+
+		// Tell TaskRunner to restart due to failure
+		const failure = true
+		c.task.Restart("healthcheck", fmt.Sprintf("check %q unhealthy", c.checkName), failure)
+
+		// Reset grace time to grace + restart.delay
+		c.graceUntil = now.Add(c.grace + c.restartDelay)
+		c.unhealthyStart = time.Time{}
+	}
+}
+
+// checkWatcher watches Consul checks and restarts tasks when they're
+// unhealthy.
+type checkWatcher struct {
+	consul ChecksAPI
+
+	// pollFreq is how often to poll the checks API and defaults to
+	// defaultPollFreq
+	pollFreq time.Duration
+
+	// watchCh is how watches (and removals) are sent to the main watching loop
+	watchCh chan *checkRestart
+
+	// done is closed when Run has exited
+	done chan struct{}
+
+	// lastErr is true if the last Consul call failed. It is used to
+	// squelch repeated error messages.
+	lastErr bool
+
+	logger *log.Logger
+}
+
+// newCheckWatcher creates a new checkWatcher but does not call its Run method.
+func newCheckWatcher(logger *log.Logger, consul ChecksAPI) *checkWatcher {
+	return &checkWatcher{
+		consul:   consul,
+		pollFreq: defaultPollFreq,
+		watchCh:  make(chan *checkRestart, 8),
+		done:     make(chan struct{}),
+		logger:   logger,
+	}
+}
+
+// Run the main Consul checks watching loop to restart tasks when their checks
+// fail. Blocks until context is canceled.
+func (w *checkWatcher) Run(ctx context.Context) {
+	defer close(w.done)
+
+	// map of check IDs to their metadata
+	checks := map[string]*checkRestart{}
+
+	// timer for check polling
+	checkTimer := time.NewTimer(0)
+	defer checkTimer.Stop() // ensure timer is never leaked
+
+	// Main watch loop
+	for {
+		// Don't start watching until we actually have checks that
+		// trigger restarts.
+		for len(checks) == 0 {
```

Comment:
Lets just create a single select block. We can start the timer when we get a watch: https://github.com/hashicorp/nomad/blob/master/client/alloc_runner_health_watcher.go#L447